# HackerNews Post Analysis
## Ask HN versus Show HN post comment frequency
## What is the best time to post

The data set takes a 20,000 rows of post data from the HackerNews website. The goal of this project is to come to analyze two scenarios:

* Do posts who's titles start with "Ask HN" or "Show HN" receieve more comments on average?
* Do posts created at certain times receive more comments on average? 

Open the hacker news csv file and print the first five rows. 

In [107]:
Opened_file = open('hacker_news.csv')
from csv import reader
read_file = reader(Opened_file)
hn = list(read_file)

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


Extract the first row of the data and assign it to the variable **headers**
Remove the first row from the data set **hn**
Display **headers**
Display the first five rows of **hn** to verify that the header row was removed

In [108]:
headers = hn[0]
hn = hn[1:]

print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


Creating 3 empty lists called **ask_posts**, **show_posts**, and **other_posts**

Loop through each row in **hn**:
* assign the title in each row to a variable named **title**
* if the lowercase version of **title** starts with **ask hn**, append teh row to **ask_posts**
* else, iff the lowercase version of **title** starts with **show hn**, append the row to **show_posts**
* else append to **other_posts**

Check the number of posts in **ask_posts**, **show_posts**, and **other_posts**

In [109]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


Find the total number of comments in ask posts and assign it to **total_ask_comments**

Loop through the ask posts to:
* add the value to the **total_ask_comments** variable set earlier
* note: will have to convert the value to an integer to calculate the sum of the comments

Compute the avg number of comments on ask posts and assign it to **avg_ask_comments**

Print **avg_ask_comments**

Do same as above, for show posts

**Findings in Markdown cell under code**

In [110]:
total_ask_comments = 0

for post in ask_posts:
    total_ask_comments += int(post[4])
    
avg_ask_comments = total_ask_comments / len(ask_posts)

print(avg_ask_comments)

14.038417431192661


In [111]:
total_show_comments = 0

for post in show_posts:
    total_show_comments += int(post[4])
    
    
avg_show_comments = total_show_comments / len(show_posts)

print(avg_show_comments)

10.31669535283993


**FINDINGS**

Ask posts average more comments per post than show posts by 4 comments/post. 

14.04 : 10.32

## Calculating the amount of ask posts and comments by hour created

import the **datetime** module as **dt**

create an empty list of lists and assign it to **result_list**

iterate over **ask_posts** and append to **result_list** a list with 2 elements:
* first element is the column **created_at**
* second element is the number of comments on the post (needs to be converted to an integer)

Create 2 empty dictionaries called **counts_by_hour** and **comments_by_hour**

Loop through each row of **result_list**
* extract the hour from teh date, which is the first element of the row
* use the **datetime.strptime()** method to parse the date and create a datetime object
* use the string we want to parse as the first argument and a string that specifies the format as the second argument

In [112]:
import datetime as dt

result_list = []

for row in ask_posts:
    result_list.append([row[6], int(row[4])]) # creating list of lists
    
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    date = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M") # parse date and create datetime object
    hour = dt.datetime.strftime(date, "%H") # extract hour from date
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = row[1]
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += row[1]
    
    

## Calculating the Average Number of Comments/Post Created During Each Hour of the Day

In [113]:
avg_by_hour = []

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr] / counts_by_hour[hr]])
    
avg_by_hour

[['15', 38.5948275862069],
 ['09', 5.5777777777777775],
 ['10', 13.440677966101696],
 ['05', 10.08695652173913],
 ['01', 11.383333333333333],
 ['18', 13.20183486238532],
 ['20', 21.525],
 ['03', 7.796296296296297],
 ['07', 7.852941176470588],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['04', 7.170212765957447],
 ['22', 6.746478873239437],
 ['12', 9.41095890410959],
 ['19', 10.8],
 ['21', 16.009174311926607],
 ['06', 9.022727272727273],
 ['13', 14.741176470588234],
 ['08', 10.25],
 ['02', 23.810344827586206],
 ['00', 8.127272727272727],
 ['23', 7.985294117647059],
 ['17', 11.46],
 ['11', 11.051724137931034]]

## Sorting the list of lists created above by hour

In [114]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

[[38.5948275862069, '15'], [5.5777777777777775, '09'], [13.440677966101696, '10'], [10.08695652173913, '05'], [11.383333333333333, '01'], [13.20183486238532, '18'], [21.525, '20'], [7.796296296296297, '03'], [7.852941176470588, '07'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.170212765957447, '04'], [6.746478873239437, '22'], [9.41095890410959, '12'], [10.8, '19'], [16.009174311926607, '21'], [9.022727272727273, '06'], [14.741176470588234, '13'], [10.25, '08'], [23.810344827586206, '02'], [8.127272727272727, '00'], [7.985294117647059, '23'], [11.46, '17'], [11.051724137931034, '11']]


In [115]:
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
print("Top 5 Hours For Ask Posts Comments")

for row in sorted_swap:
    time = dt.time(hour=int(row[1])).strftime("%H:%M")
    print("{time}:{avg:.2f} average comments per post".format(time=time, avg=row[0]))

Top 5 Hours For Ask Posts Comments
15:00:38.59 average comments per post
02:00:23.81 average comments per post
20:00:21.52 average comments per post
16:00:16.80 average comments per post
21:00:16.01 average comments per post
13:00:14.74 average comments per post
10:00:13.44 average comments per post
14:00:13.23 average comments per post
18:00:13.20 average comments per post
17:00:11.46 average comments per post
01:00:11.38 average comments per post
11:00:11.05 average comments per post
19:00:10.80 average comments per post
08:00:10.25 average comments per post
05:00:10.09 average comments per post
12:00:9.41 average comments per post
06:00:9.02 average comments per post
00:00:8.13 average comments per post
23:00:7.99 average comments per post
07:00:7.85 average comments per post
03:00:7.80 average comments per post
04:00:7.17 average comments per post
22:00:6.75 average comments per post
09:00:5.58 average comments per post
